src/java/main/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -87,6 +87,12 @@ public void close() {
             LOG.debug("close called for sessionid:0x"
                     + Long.toHexString(sessionId));
         }
+
+        // ZOOKEEPER-2743:
+        // Always unregister connection upon close to prevent
+        // connection bean leak under certain race conditions.
+        factory.unregisterConnection(this);
```

Comment:
Agreed Michael. I've noticed `NIOServerCnxnFactory#removeCnxn()` execution path, we should correct this part as well, like you mentioned in the PR comment. Could you please raise a new jira to track the  changes.

+1 the netty code changes looks good to me.